xarray/backends/common.py

```diff
@@ -374,14 +436,19 @@ class DataStorePickleMixin(object):
     def __getstate__(self):
         state = self.__dict__.copy()
         del state['ds']
+        state['_isopen'] = False
         if self._mode == 'w':
             # file has already been created, don't override when restoring
             state['_mode'] = 'a'
         return state
 
     def __setstate__(self, state):
         self.__dict__.update(state)
-        self.ds = self._opener(mode=self._mode)
+        if self._autoclose:
+            with self.ensure_open(True):
+                self.ds = self._opener(mode=self._mode)
+        else:
+            self.ds = self._opener(mode=self._mode)
```

Comment:
This isn't pretty but we don't really want to open the dataset when un-pickling. This actually opens then closes. Maybe this needs a special lock statement?

Reply:
I think we could only restore `self.ds` if not using autoclose. There's no point in restoring the dataset only to immediately close it, given that we assign `self.ds` in `ensure_open()` anyways if using autoclose.

So this could just be:
```python
if not self._autoclose:
    self.ds = self._opener(mode=self._mode)
```

Reply:
Right. I ended up doing something a bit different by writing `ds` as a property. I'm not sure I like it but it seems to work.